torch/csrc/autograd/variable.cpp

```diff
@@ -1,102 +1,140 @@
-#include "Python.h"
 #include "torch/csrc/autograd/variable.h"
 
 #include "torch/csrc/assertions.h"
-#include "torch/csrc/autograd/generated/VariableType.h"
-#include "torch/csrc/autograd/generated/Functions.h"
+#include "torch/csrc/autograd/edge.h"
+#include "torch/csrc/autograd/function.h"
 #include "torch/csrc/autograd/functions/accumulate_grad.h"
 #include "torch/csrc/autograd/functions/tensor.h"
-
+#include "torch/csrc/autograd/generated/Functions.h"
+#include "torch/csrc/autograd/generated/VariableType.h"
+#include "torch/csrc/autograd/variable_version.h"
+#include "torch/csrc/jit/tracer_state.h"
+#include "torch/csrc/utils/auto_unique_ptr.h"
+
+#include <ATen/Scalar.h>
+#include <ATen/ScalarType.h>
+#include <ATen/Storage.h>
+#include <ATen/Tensor.h>
+#include <ATen/TensorImpl.h>
+#include <ATen/Type.h>
+
+#include <list>
 #include <memory>
-
-using namespace at;
+#include <mutex>
+#include <stdexcept>
+#include <string>
+#include <vector>
 
 namespace torch { namespace autograd {
-
-Variable make_variable(at::Tensor data, std::shared_ptr<Function> grad_fn) {
-  // TODO: If you ever want to support returning an undefined tensor from
-  // a function, you'll have to uncomment the line below.  Not sure if
-  // we actually want to support this.
-  // if (!data.defined()) return Variable();
-  TORCH_ASSERT(grad_fn);
-  int output_nr = grad_fn->num_inputs++;
-  return make_variable(std::move(data), output_nr, std::move(grad_fn));
-}
-
-VariableImpl::VariableImpl(Tensor data_, bool requires_grad, int output_nr, std::shared_ptr<Function> grad_fn)
-  : TensorImpl(VariableType::getType(data_))
-  , data(std::move(data_))
-  , grad()
-  , _grad_fn(std::move(grad_fn))
-  , version_counter()
-  , _requires_grad(requires_grad)
-  , is_view(false)
-  , output_nr(output_nr)
-  , pyobj(nullptr) {
-  TORCH_ASSERTM(!_grad_fn || !_requires_grad, "_requires_grad should be false if grad_fn is set");
+namespace {
+at::Tensor handle_scalars(at::Tensor& data) {
+#ifndef WITH_SCALARS
+  if (data.dim() == 0) {
+    // Don't expose 0-dim tensors to Variable API.
+    return data.as_strided_({1}, {1});
+  }
+#endif
+  return data;
+}
+} // namespace
+
+Variable::Impl::Impl(at::Tensor data_, bool requires_grad_, Edge gradient_edge)
+    : TensorImpl(VariableType::getType(data_)),
+      data(std::move(data_)),
+      grad_fn(std::move(gradient_edge.function)),
+      requires_grad(requires_grad_),
+      output_nr(gradient_edge.input_nr) {
```

Comment:
I don't feel like having initialization of some fields specified in the header, while others mentioned here is a good idea. It makes it very easy to forget to initialize a certain fields, especially that those structs are very complicated.